ir/shared/src/main/scala/org/scalajs/ir/Trees.scala

```diff
@@ -283,6 +283,103 @@ object Trees {
     val tpe = AnyType
   }
 
+  /** Apply a typed closure
+   *
+   *  The given `fun` must have a closure type.
+   *
+   *  The arguments' types must match (be subtypes of) the parameter types of
+   *  the closure type.
+   *
+   *  The `tpe` of this node is the result type of the closure type, or
+   *  `nothing` if the latter is `nothing`.
+   *
+   *  Evaluation steps are as follows:
+   *
+   *  1. Let `funV` be the result of evaluating `fun`.
+   *  2. If `funV` is `nullClosure`, trigger an NPE undefined behavior.
+   *  3. Let `argsV` be the result of evaluating `args`, in order.
+   *  4. Invoke `funV` with arguments `argsV`, and return the result.
+   */
+  sealed case class ApplyTypedClosure(flags: ApplyFlags, fun: Tree, args: List[Tree])(
+      implicit val pos: Position)
+      extends Tree {
+
+    val tpe: Type = fun.tpe match {
+      case ClosureType(_, resultType, _) => resultType
+      case NothingType                   => NothingType
+      case _                             => NothingType // never a valid tree
+    }
+  }
+
+  /** New lambda instance of a SAM class.
+   *
+   *  The `fun` must have a non-nullable `ClosureType` whose signature matches
+   *  the signature of the given `method`.
+   *
+   *  Functionally, a `NewLambda` is equivalent to an instance of an anonymous
+   *  class with the following shape:
+   *
+   *  {{{
+   *  val funV: ((...Ts) => R)! = fun;
+   *  (new superClass with interfaces {
+   *    def <this>() = this.superClass::<init>()
+   *    def method(...args: Ts): R = funV(...args)
+   *  }): tpe
+   *  }}}
+   *
+   *  where `superClass`, `interfaces` and `method` are taken from the
+   *  `descriptor`.
+   *
+   *  There must exist `C ∈ { superClass } ∪ interfaces` such that `C!` is a
+   *  subtype of `tpe`.
```

Comment:
I find this hard to understand. What it expresses is the (obvious) requirement that `tpe >: superClass with interfaces` by taking into account properties of the IR type system (notably that there are no intersection types).

I feel simply specifying `tpe >: superClass with interfaces` and additionally explaining the consequence might be easier to understand.